doc/reference.md

```diff
@@ -0,0 +1,509 @@
+
+### Array
+
+Arrays in Jsonnet are finite-length sequences of arbitrary values.
+Values of different types can be mixed in an array. Individual array elements are lazy, meaning that evaluating an array does not cause evaluation of all arguments.
+
+There is no separate tuple type in Jsonnet. Arrays are used in contexts where tuples would be natural in other languages, for example for returning multiple values from a function.
+
+The simplest way of creating an array is *an array literal*. It is simply a comma separated list of elements, such as `[1, 2, "foo", 2 + 2]`.
+
+The most flexible way of creating an array is `std.makeArray(sz, func)`. It takes the size of the array to construct and a function which takes an index `i` and returns `i`-th element. It is possible to build all other array functionality using this function. In practice, using more specialized functions is usually (but not always) more handy and results in a more efficient program.
+
+Jsonnet also offers *array comprehensions* which provide an elegant and easy to use syntax for mapping, filtering and taking Cartesian products of arrays. Array comprehensions can be "mechanically" translated to a series of `std.flatMap` calls and simple conditionals. For example, the following programs are equivalent:
+
+```
+[ 
+  [x * 2, y]
+  for x in [1, 2, 3, 4, 5]
+  for y in [1, 2, 3]
+  if x % 2 == 0
+]
+```
+
+```
+std.flatMap(
+  function(x) std.flatMap(
+    function(y) if x % 2 == 0 then [[x * 2, y]] else [],
+    [1, 2, 3]
+  ),
+  [1, 2, 3, 4, 5]
+)
+```
+
+Arrays can be concatenated using the operator `+`. Arrays `a` and `b` are equal if they have equal length and for all indexes `i`, `a[i] == b[i]`. The comparison of arrays is "lexicographic", so array `a` is smaller than `b` if `a[i] < b[i]` for some `i` and for all `j < i`, `a[i] == b[i]`. 
+
+### Object
+
+In the simplest case Jsonnet object is a mapping from string keys to arbitrary values.
+
+```
+{
+  "foo": 1,
+  "bar": {
+    "arr": [1, 2, 3],
+    "number": 10 + 7,
+  }
+}
+```
+
+Jsonnet objects can be indexed either using `.` and an identifier or `[]` with an arbitrary expression.
+
+```
+local obj = {
+  "foo": 1,
+  "bar": {
+    "arr": [1, 2, 3],
+    "number": 10 + 7,
+  }
+};
+[
+  obj.foo,
+  obj["foo"],
+  obj["f" + "oo"]
+]
+```
+
+### Inheritance
+
+Jsonnet objects allow inheritance in the OOP sense, even though there are no classes or declarations. The inheritance is realized as an operation `+` which can be applied to any two objects. This might be surprising, because in mainstream languages the inheritance hierarchy is static.
+
+For objects which are simple key-value mappings, inheritance is the same as replacing the respective fields from the first object with fields from the second object.
+
+```
+{
+  a: 1
+  b: 2,
+}
++
+{
+  a: 3
+}
+```
+
+It is possible for one field to refer to other ones using `self` and it is possible to refer to the fields of the base object using `super`.
```

Comment:
I've applied the suggestion and then rephrased it more, to avoid referring to "the base object", because there is really no such thing (there are just layers in the newly created object).